File: iOS/Notes/Notes/Application/Root/State/RootState.swift
//  Created by Denis Malykh on 21.03.2023.

import Foundation

struct RootState: State {

    enum Actions: Action {
        case switchWorkMode(mode: WorkMode)
    }

    let currentWorkMode: WorkMode

    static func provideReducer() -> Store<RootState, RootState.Actions>.Reducer {
        { _, action in
            switch action {
            case let .switchWorkMode(mode):
                return RootState(currentWorkMode: mode)
            }
        }
    }
}

File: Android/app/src/main/java/ru/mrdekk/notes/app/root/state/RootState.kt
package ru.mrdekk.notes.app.root.state

import ru.mrdekk.notes.app.root.model.WorkMode
import ru.mrdekk.notes.generic.arch.Reducer
import ru.mrdekk.notes.generic.arch.action.Action
import ru.mrdekk.notes.generic.arch.state.State

class RootState(
    val currentWorkMode: WorkMode
): State {
    sealed interface Actions: Action {
        data class SwitchWorkMode(val mode: WorkMode): Actions
    }

    companion object {
        fun provideReducer(): Reducer<RootState, Actions> = { _, action ->
            when (action) {
                is Actions.SwitchWorkMode -> {
                    RootState(currentWorkMode = action.mode)
                }
            }
        }
    }
}
